torchgeo/trainers/classification.py

```diff
@@ -238,6 +238,20 @@ def test_epoch_end(self, outputs: Any) -> None:
         self.log_dict(self.test_metrics.compute())
         self.test_metrics.reset()
 
+    def predict_step(self, *args: Any, **kwargs: Any) -> Tensor:
+        """Compute and return the predictions.
+
+        Args:
+            batch: the output of your DataLoader
+
+        Returns:
+            predicted softmax probabilities
+        """
+        batch = args[0]
+        x = batch["image"]
+        y_hat: Tensor = self(x).softmax(dim=-1)
+        return y_hat
```

Comment:
Why is the predict step so different from the train/val/test steps? Shouldn't it look something like:
```suggestion
        y_hat = self.forward(x)
        y_hat_hard = y_hat.argmax(dim=1)
        return y_hat_hard
```

Reply:
Because you will frequently want the per-class probabilities instead of just the top class

Reply:
And that's not true for train/val/test?

Reply:
The train/val/test steps aren't meant to return data to a user. The operations that happen are only meant to perform the forward pass and compute metrics (which is why the argmax op is needed). The predict step needs to return the raw softmax outputs because we don't want to assume what a user wants to do with the output.

Reply:
Do we need `yhat: Tensor` or can we just use `yhat`?

Reply:
I had to add that to avoid mypy complaining.

Reply:
Why was it complaining? If you use `self.forward(x)` instead does it stop complaining? What if you break it into multiple lines like we do in other steps?

Reply:
(type hint warnings are not something to be ignored, they're something to be fixed)

Reply:
- The recommended way to run inference is with `self(x)` not `self.forward(x)` as `self(x)` calls hooks while `self.forward(x)` silently doesn't.
- `y_hat` is a Tensor so I don't see the problem. Breaking something into multiple lines to get `mypy` to stop complaining is ridiculous overhead.

Reply:
Should we change other methods to replace `self.forward(x)` with `self(x)`? Just want to make sure we're consistent.

Reply:
Yes we should change. We shouldn't be calling forward anywhere.

Reply:
P.S. I feel like the `yhat: Tensor` thing is due to missing type hints for softmax but I couldn't figure out exactly where I need to add those. PyTorch's source code is a maze.

Reply:
I think I was able to fix it by calling it as a function e.g. torch.softmax(tensor) instead of calling it as a method e.g. tensor.softmax()